File: Fortran/01_Recap/03_Reduction/maxloc.f90
PROGRAM reduce

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100

  INTEGER :: rank, recv
  INTEGER :: nproc, ierr
  INTEGER, DIMENSION(:), ALLOCATABLE :: seed
  INTEGER :: rand_size, time, rank_recv
  REAL :: randval
   :: maxloc_data_out, maxloc_data_in

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  CALL RANDOM_SEED(size=rand_size)
  ALLOCATE(seed(1:rand_size))
  CALL SYSTEM_CLOCK(time)
  seed = time + rank
  CALL RANDOM_SEED(put=seed)
  DEALLOCATE(seed)
  CALL RANDOM_NUMBER(randval)
  randval = randval * 100

  !Note that unlike in C, we're using an array of reals. The second one contains
  !an integer. This is necessary for compatability with F77 which has no structs
  maxloc_data_out(1) = randval
  maxloc_data_out(2) = rank

  !Note that type is MPI_2REAL. That's two reals packed together
  !Also
  !MPI_2DOUBLE_PRECISION - Two double precisions packed together
  !MPI_2INTEGER - Two integers packed together
  CALL MPI_Reduce(maxloc_data_out, maxloc_data_in, 1, MPI_2REAL, MPI_MAXLOC, &
      0, MPI_COMM_WORLD, ierr)
  rank_recv = maxloc_data_in(2)

  PRINT *, "On rank ", rank, " random number was ", randval

  IF (rank == 0) THEN
  PRINT *, "MPI_Reduce with MPI_MAXLOC gives maximum random value as ", &
      maxloc_data_in(1), " on rank ", rank_recv
  END IF

  CALL MPI_Finalize(ierr)

END PROGRAM reduce

File: C/01_Recap/03_Reduction/maxloc.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <mpi.h>

#define TAG 100

int main(int argc, char ** argv)
{

  int rank, nproc, recv, randval;
  time_t t;

  struct {
  } maxloc_data_out, maxloc_data_in;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  //Get random number between 0 and 99. + rank so different number on each
  srand((unsigned) time(&t)+ rank);
  randval = rand() % 100;

  maxloc_data_out.value = randval;
  maxloc_data_out.rank = rank; //Have to put rank in here manually

  //Note that type is MPI_2INT. That's two integers packed together
  //Also
  /*MPI_FLOAT_INT - Float and Int
    MPI_DOUBLE_INT - Double and Int
    MPI_LONG_INT - Long and Int
    MPI_SHORT_INT - Short and Int
    MPI_LONG_DOUBLE_INT - Long Double and Int
  */
  MPI_Reduce(&maxloc_data_out, &maxloc_data_in, 1, MPI_2INT, MPI_MAXLOC, 0,
      MPI_COMM_WORLD);

  printf("On rank %3d, random number was %3d\n", rank, randval);

  if (rank == 0) {
  printf("MPI_Reduce with MPI_MAXLOC gives maximum random value as %3d on rank %3d\n"
      , maxloc_data_in.value, maxloc_data_in.rank);
  }

  MPI_Finalize();

}
